src/components/coin.ts

```diff
@@ -72,3 +91,75 @@ export const adjustCoinBalanceByUserId = async (userId: string, amount: number):
     amount
   );
 };
+
+/*
+  If (user, bonusType) doesn't exist, create row with current time as this bonusType log.
+  Otherwise, update last_granted to CURRENT_TIMESTAMP.
+*/
+export const updateUserBonusTableByUserId = async (userId: string, bonusType: BonusType): Promise<void> => {
+  const db = await openDB();
+  await db.run(
+    `
+    INSERT INTO user_coin_bonus (user_id, bonus_type, last_granted) VALUES (?, ?, CURRENT_TIMESTAMP)
+    ON CONFLICT(user_id, bonus_type)
+    DO UPDATE SET last_granted = CURRENT_TIMESTAMP`,
+    userId,
+    bonusType
+  );
+};
+
+/*
+  Get the time of the latest bonus applied to a user based on type
+*/
+export const latestBonusByUserId = async (userId: string, type: BonusType): Promise<UserCoinBonus | undefined> => {
+  const db = await openDB();
+  const res: UserCoinBonus | undefined = await db.get(
+    `SELECT last_granted FROM user_coin_bonus WHERE user_id = ? AND bonus_type = ?`,
+    userId,
+    type
+  );
+  return res;
+};
+
+/*
+  Adjusts balance by bonus amount if appropriate cooldown has passed.
+  Returns bool. 
+  True initiates the break in applyBonusByUserId, preventing multiple bonuses for a single msg.
```

Comment:
IMO this is breaking the separation of concern with `applyBonusByUserId`, this function shouldn't worry about how the other function will use the value of this function.